Add bufferization support for UKernel ops.

MaheshRavishankar opened the pull request:
This adds an `ExternalModel` implementation for the `BufferizationInterface` that derives from
`bufferization::DstBufferizableOpInterfaceExternalModel`, allowing the OneShot Bufferization to pick up these ops.

Commit:
Add bufferization support for UKernel ops.

This adds an `ExternalModel` implementation for the
`BufferizationInterface` that derives from
`bufferization::DstBufferizableOpInterfaceExternalModel`, allowing the
OneShot Bufferization to pick up these ops.

Commit:
Fix bazel build.

matthias-springer commented on compiler/src/iree/compiler/Codegen/Dialect/UKernelOps.cpp:
```diff
@@ -327,8 +329,81 @@ FailureOr<func::CallOp> UKernelMmt4DOp::lowerToFunctionCall(
+                                          bufferOpOperands, op->getAttrs());
+    SmallVector<Value> dpsInits =
+        cast<DestinationStyleOpInterface>(bufferOp.getOperation())
+            .getDpsInitOperands();
```
Does `getDpsInitOperands` (or any DPS interface method) work on ops with memref semantics? Maybe query the inits from the original unbufferized op.

MaheshRavishankar replied:
It should work on `memref` operands as well. Do you know that it is not indeed the fact.

bjacob committed:
Add missing `template` keyword (fix GCC build)

 https://en.cppreference.com/w/cpp/language/dependent_name -> paragraph "The template disambiguator for dependent names"

bjacob reviewed (approved):
I committed directly a tiny fix that should resolve the CI errors with GCC.
